File: exercises/2-intermediate/config/game-controller/main.go
package main

import (
    "context"
    "flag"
    "fmt"
    "os"
    "path/filepath"
    
    appsv1 "k8s.io/api/apps/v1"
    corev1 "k8s.io/api/core/v1"
    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
    "k8s.io/apimachinery/pkg/util/intstr"
    "k8s.io/client-go/kubernetes"
    "k8s.io/client-go/tools/clientcmd"
    "k8s.io/client-go/util/homedir"
)

// Game represents our custom resource
type Game struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    Spec             GameSpec `json:"spec"`
}

// GameSpec defines our game server configuration
type GameSpec struct {
    GameName string `json:"gameName"`
    Players  int32  `json:"players"`
    Port     int32  `json:"port"`
}

func main() {
    // Setup kubernetes config
    kubeconfig := flag.String("kubeconfig", filepath.Join(homedir.HomeDir(), ".kube", "config"), "")
    flag.Parse()

    config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
    if err != nil {
        fmt.Printf("Error building config: %v\n", err)
        os.Exit(1)
    }

    clientset, err := kubernetes.NewForConfig(config)
    if err != nil {
        fmt.Printf("Error creating clientset: %v\n", err)
        os.Exit(1)
    }

    // Example game server
    game := &Game{
        ObjectMeta: metav1.ObjectMeta{
            Name: "minecraft-server",
            Namespace: "default",
        },
        Spec: GameSpec{
            GameName: "minecraft",
            Players: 20,
            Port: 25565,
        },
    }

    // Create deployment for the game server
    deployment := createGameDeployment(game)
    _, err = clientset.AppsV1().Deployments(game.Namespace).Create(context.TODO(), deployment, metav1.CreateOptions{})
    if err != nil {
        fmt.Printf("Error creating deployment: %v\n", err)
    }

    // Create service to expose the game server
    service := createGameService(game)
    _, err = clientset.CoreV1().Services(game.Namespace).Create(context.TODO(), service, metav1.CreateOptions{})
    if err != nil {
        fmt.Printf("Error creating service: %v\n", err)
    }
}

func createGameDeployment(game *Game) *appsv1.Deployment {
    replicas := int32(1)
    return &appsv1.Deployment{
        ObjectMeta: metav1.ObjectMeta{
            Name: game.Name,
            Namespace: game.Namespace,
        },
        Spec: appsv1.DeploymentSpec{
            Replicas: &replicas,
            Selector: &metav1.LabelSelector{
                MatchLabels: map[string]string{
                    "game": game.Name,
                },
            },
            Template: corev1.PodTemplateSpec{
                ObjectMeta: metav1.ObjectMeta{
                    Labels: map[string]string{
                        "game": game.Name,
                    },
                },
                Spec: corev1.PodSpec{
                    Containers: []corev1.Container{
                        {
                            Name:  game.Spec.GameName,
                            Image: fmt.Sprintf("%s:latest", game.Spec.GameName),
                            Ports: []corev1.ContainerPort{
                                {
                                    ContainerPort: game.Spec.Port,
                                },
                            },
                            Env: []corev1.EnvVar{
                                {
                                    Name: "MAX_PLAYERS",
                                    Value: fmt.Sprintf("%d", game.Spec.Players),
                                },
                            },
                        },
                    },
                },
            },
        },
    }
}

func createGameService(game *Game) *corev1.Service {
    return &corev1.Service{
        ObjectMeta: metav1.ObjectMeta{
            Name: game.Name,
            Namespace: game.Namespace,
        },
        Spec: corev1.ServiceSpec{
            Type: corev1.ServiceTypeNodePort,
            Selector: map[string]string{
                "game": game.Name,
            },
            Ports: []corev1.ServicePort{
                {
                    Port: game.Spec.Port,
                    TargetPort: intstr.FromInt(int(game.Spec.Port)),
                },
            },
        },
    }
}

File: exercises/2-intermediate/config/game-controller/main.py
from kubernetes import client, config
from kubernetes.client import V1Deployment, V1Service
from kubernetes.client.rest import ApiException
from dataclasses import dataclass
import os
import argparse
from pathlib import Path

@dataclass
class GameSpec:
    game_name: str
    players: int
    port: int

@dataclass
class Game:
    name: str
    namespace: str
    spec: GameSpec

def create_game_deployment(game: Game) -> V1Deployment:
    return client.V1Deployment(
        metadata=client.V1ObjectMeta(
            name=game.name,
            namespace=game.namespace
        ),
        spec=client.V1DeploymentSpec(
            replicas=1,
            selector=client.V1LabelSelector(
                match_labels={"game": game.name}
            ),
            template=client.V1PodTemplateSpec(
                metadata=client.V1ObjectMeta(
                    labels={"game": game.name}
                ),
                spec=client.V1PodSpec(
                    containers=[
                        client.V1Container(
                            name=game.spec.game_name,
                            image=f"{game.spec.game_name}:latest",
                            ports=[
                                client.V1ContainerPort(
                                    container_port=game.spec.port
                                )
                            ],
                            env=[
                                client.V1EnvVar(
                                    name="MAX_PLAYERS",
                                    value=str(game.spec.players)
                                )
                            ]
                        )
                    ]
                )
            )
        )
    )

def create_game_service(game: Game) -> V1Service:
    return client.V1Service(
        metadata=client.V1ObjectMeta(
            name=game.name,
            namespace=game.namespace
        ),
        spec=client.V1ServiceSpec(
            type="NodePort",
            selector={"game": game.name},
            ports=[
                client.V1ServicePort(
                    port=game.spec.port,
                    target_port=game.spec.port
                )
            ]
        )
    )

def main():
    # Parse arguments
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--kubeconfig",
        default=os.path.join(str(Path.home()), ".kube", "config"),
        help="Path to kubeconfig file"
    )
    args = parser.parse_args()

    # Load kubernetes configuration
    try:
        config.load_kube_config(args.kubeconfig)
    except Exception as e:
        print(f"Error loading kubeconfig: {e}")
        exit(1)

    # Create API clients
    apps_v1 = client.AppsV1Api()
    core_v1 = client.CoreV1Api()

    # Create game instance
    game = Game(
        name="minecraft-server",
        namespace="default",
        spec=GameSpec(
            game_name="minecraft",
            players=20,
            port=25565
        )
    )

    # Create deployment
    try:
        deployment = create_game_deployment(game)
        apps_v1.create_namespaced_deployment(
            namespace=game.namespace,
            body=deployment
        )
        print(f"Created deployment: {game.name}")
    except ApiException as e:
        print(f"Error creating deployment: {e}")

    # Create service
    try:
        service = create_game_service(game)
        core_v1.create_namespaced_service(
            namespace=game.namespace,
            body=service
        )
        print(f"Created service: {game.name}")
    except ApiException as e:
        print(f"Error creating service: {e}")

if __name__ == "__main__":
    main()